app/plugin.go

```diff
@@ -4,356 +4,133 @@
-		}
-
-		if activate && enabled {
-			a.setPluginActive(plugin, activate)
-		} else if !activate {
-			a.setPluginActive(plugin, activate)
-		}
-	}
-
-	if err := a.notifyPluginStatusesChanged(); err != nil {
-		mlog.Error("failed to notify plugin status changed", mlog.Err(err))
-	}
-}
-
-func (a *App) setPluginActiveById(id string, activate bool) {
-	plugins, err := a.PluginEnv.Plugins()
-	if err != nil {
-		mlog.Error(fmt.Sprintf("Cannot setPluginActiveById(%t)", activate), mlog.String("plugin_id", id), mlog.Err(err))
-		return
-	}
+	config := a.Config().PluginSettings
 
-	for _, plugin := range plugins {
-		if plugin.Manifest != nil && plugin.Manifest.Id == id {
-			a.setPluginActive(plugin, activate)
+	if *config.Enable {
+		availablePlugins, err := a.Plugins.Available()
+		if err != nil {
+			a.Log.Error("Unable to get available plugins", mlog.Err(err))
+			return
 		}
-	}
-}
-
-func (a *App) setPluginActive(plugin *model.BundleInfo, activate bool) {
-	if plugin.Manifest == nil {
-		return
-	}
-
-	id := plugin.Manifest.Id
-
-	active := a.PluginEnv.IsPluginActive(id)
 
-	if activate {
-		if !active {
-			if err := a.activatePlugin(plugin.Manifest); err != nil {
-				mlog.Error("Plugin failed to activate", mlog.String("plugin_id", plugin.Manifest.Id), mlog.String("err", err.DetailedError))
+		// Deactivate any plugins that have been disabled.
+		for _, plugin := range a.Plugins.Active() {
+			// Determine if plugin is enabed
+			pluginId := plugin.Manifest.Id
+			pluginEnabled := false
+			if state, ok := config.PluginStates[pluginId]; ok {
+				pluginEnabled = state.Enable
 			}
-		}
 
-	} else if !activate {
-		if active {
-			if err := a.deactivatePlugin(plugin.Manifest); err != nil {
-				mlog.Error("Plugin failed to deactivate", mlog.String("plugin_id", plugin.Manifest.Id), mlog.String("err", err.DetailedError))
-			}
-		} else {
-			if err := a.setPluginStatusState(plugin.Manifest.Id, model.PluginStateNotRunning); err != nil {
-				mlog.Error("Plugin status state failed to update", mlog.String("plugin_id", plugin.Manifest.Id), mlog.String("err", err.Error()))
+			// If it's not enabled we need to stop it
+			if !pluginEnabled {
+				a.Plugins.Deactivate(pluginId)
 			}
 		}
-	}
-}
-
-func (a *App) activatePlugin(manifest *model.Manifest) *model.AppError {
-	mlog.Debug("Activating plugin", mlog.String("plugin_id", manifest.Id))
-
-	if err := a.setPluginStatusState(manifest.Id, model.PluginStateStarting); err != nil {
-		return model.NewAppError("activatePlugin", "app.plugin.set_plugin_status_state.app_error", nil, err.Error(), http.StatusInternalServerError)
-	}
 
-	onError := func(err error) {
-		mlog.Debug("Plugin failed to stay running", mlog.String("plugin_id", manifest.Id), mlog.Err(err))
+		// Activate any plugins that have been enabed
+		for _, plugin := range availablePlugins {
+			if plugin.Manifest == nil {
+				plugin.WrapLogger(a.Log).Error("Plugin manifest could not be loaded", mlog.Err(plugin.ManifestError))
```

Comment:
I guess this will call an error every time we `SyncPluginsActiveState`? Is that spammy?

Reply:
It's a bit spammy, but because of the way we do plugins I think this is the best place to put it. Plugins not parsing correctly should be rare and immediately fixed anyway.